spec/engine_spec.rb

```diff
@@ -0,0 +1,271 @@
+    end
+  end
+
+  describe "#process" do
+    let(:block) { proc { |f| "foo" } }
+
+    context "for Message::WorkerStarded" do
+      let(:mesg) { Ronin::Recon::Message::WorkerStarted.new(worker) }
+
+      it "must increment #worker_task_count" do
+        Async { subject.process(mesg) }
+
+        expect(subject.instance_variable_get(:@worker_task_count)).to eq(1)
+      end
+    end
+
+    context "for Message::WorkerStopped" do
+      let(:mesg) { Ronin::Recon::Message::WorkerStopped.new(worker) }
+
+      it "must decrement #worker_taks_count" do
+        Async { subject.process(mesg) }
+
+        expect(subject.instance_variable_get(:@worker_task_count)).to eq(-1)
+      end
+    end
+
+    context "for Message::JobStarted" do
+      let(:mesg) { Ronin::Recon::Message::JobStarted.new(worker, value) }
+
+      before do
+        Async { subject.on(:job_started, &block) }
+      end
+
+      it "must call all callbacks from #job_started_callbacks" do
+        Async { subject.process(mesg) }
+      end
+    end
+
+    context "for Message::JobCompleted" do
+      let(:mesg) { Ronin::Recon::Message::JobCompleted.new(worker, value) }
+
+      before do
+        Async { subject.on(:job_completed, &block) }
+      end
+
+      it "must call all callbacks from #job_completed_callbacks" do
+        Async { subject.process(mesg) }
+      end
+    end
+
+    context "for Message::JobFailed" do
+      let(:mesg) { Ronin::Recon::Message::JobFailed.new(worker, value, "ERROR") }
+
+      before do
+        Async { subject.on(:job_failed, &block) }
+      end
+
+      it "must call all callbacks from #job_completed_callbacks" do
+        Async { subject.process(mesg) }
+      end
+    end
+
+    context "for Message::Value" do
+      let(:mesg) { Ronin::Recon::Message::Value.new(value, parent: value) }
+
+      it "must add node and edge to #graph" do
+        Async { subject.process(mesg) }
+
+        expect(subject.instance_variable_get(:@graph).nodes).to eq(Set.new([value]))
+        expect(subject.instance_variable_get(:@graph).edges).to eq({ value => Set.new([value]) })
+      end
+    end
+
+    context "for not implemented Message" do
+      let(:mesg) { Ronin::Recon::Message::Shutdown.new }
+
+      it "must raise a NotImplementedError" do
+        expect {
+          Async { subject.process(mesg) }
+        }.to raise_error(NotImplementedError)
+      end
+    end
+  end
+
+  describe "#start" do
+    before do
+      Async { subject.add_worker(TestEngine::TestWorker) }
+    end
+
+    it "must enqueue #scope values, output customer task and start all work groups" do
+      Async { subject.start }
+
+      expect(subject.values).to eq(Set.new([value]))
+    end
+  end
+
+  describe "#shutdown!" do
+    before do
+      # enqueue #output_queue
+    end
+
+    it "must enqueue Message::SHUTDOWN and dequeue #output_queue" do
+      Async { subject.shutdown! }
+
+      expect(subject.instance_variable_get(:@worker_task_count)).to eq(0)
```

Comment:
I should probably expose `worker_task_count` as `#running_workers` or something.